### PR TITLE
Export zap's internal observing logger

### PR DESCRIPTION
**Pull request by billf**
Simplify `internal/observer` and export it as `zaptest/observer`.

### CODE REVIEW

File: zapcore/tee_test.go [hunk 1]
```diff
@@ -137,10 +134,9 @@ func TestTeeEnabled(t *testing.T) {
 }
 
 func TestTeeSync(t *testing.T) {
-	tee := NewTee(
-		observer.New(InfoLevel, nil, false),
-		observer.New(WarnLevel, nil, false),
-	)
+	infoLogger, _ := observer.New(InfoLevel)
+	warnLogger, _ := observer.New(WarnLevel)
+	tee := NewTee(infoLogger, warnLogger)
```
Comment (prashantv) [inline zapcore/tee_test.go]: do we need real loggers here, can we pass `NopLoggers` or something?

File: zaptest/observer/observer.go [hunk 2]
```diff
@@ -40,13 +40,11 @@ type ObservedLogs struct {
 	logs []LoggedEntry
 }
 
-// Add appends a new observed log to the collection. It always returns nil
-// error, but does so that it can be used as an observer sink function.
-func (o *ObservedLogs) Add(log LoggedEntry) error {
+// Add appends a new observed log to the collection.
+func (o *ObservedLogs) Add(log LoggedEntry) {
```
Comment (prashantv) [inline zaptest/observer/observer.go]: does this need to be exported?

File: zapcore/sampler_test.go [hunk 3]
```diff
@@ -26,20 +26,19 @@ import (
...
-	"github.com/stretchr/testify/require"
-
 	"go.uber.org/atomic"
-	"go.uber.org/zap/internal/observer"
 	. "go.uber.org/zap/zapcore"
 	"go.uber.org/zap/zaptest"
+	"go.uber.org/zap/zaptest/observer"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
```
Comment (akshayjshah) [inline zapcore/sampler_test.go]: :+1: Thanks for the import reordering.

File: zapcore/tee_test.go (hunk 1 above)
Comment (akshayjshah) [inline zapcore/tee_test.go]: The `NopCore` is always disabled, so it's not helpful here.

File: zaptest/observer/observer.go (hunk 2 above)
Comment (akshayjshah) [inline zaptest/observer/observer.go]: Nope, I'll push a small change to un-export it.

**Pull request merged**